HDDS-9645. Recon doesn't exclude out-of-service nodes when checking for healthy containers

xBis7 opened the pull request:
## What changes were proposed in this pull request?

Datanodes that are `DECOMMISSIONING`, `DECOMMISSIONED`, `ENTERING_MAINTENANCE` or `IN_MAINTENANCE`, are considered out-of-service. SCM doesn't count their replicas as healthy and available when checking for under-replicated or over-replicated containers. 

Recon should do the same and filter out-of-service nodes because
 1. their replicas aren't actually available
 2. it needs to be consistent with SCM

This patch modifies the `ContainerHealthStatus` constructor to achieve the desired behavior.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9645

## How was this patch tested?

New unit tests were added. This patch was also tested manually. The steps for manual testing can be found on the jira's description.

Green CI builds on my fork:
https://github.com/xBis7/ozone/actions/runs/6944974103
https://github.com/xBis7/ozone/actions/runs/6945013128

xBis7 commented:
@devmadhuu Can you please take a look at this PR?

devmadhuu reviewed (approved):
@xBis7 thanks for working on this patch. Changes LGTM +1

sumitagrawl reviewed (approved):
@xBis7 Thanks for working over this, LGTM

sodonnel commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthStatus.java:
```diff
@@ -48,8 +48,12 @@ public class ContainerHealthStatus {
+        // replicas belonging to out-of-service nodes.
+        .filter(r ->
+            (!r.getDatanodeDetails().isDecommissioned() &&
+             !r.getDatanodeDetails().isMaintenance() &&
```
It is "OK" for a maintenance replica to be offline. The definition of maintenance is that one or two replicas out of 3 can be offline and the container is still considered healthy, so I am not sure if it is correct to just assume a maintenance copy is offline. If there was a scenario where all 3 of the replicas of a container were in maintenance, with no other available copies, then that would be under-replicated.

This leads to another question - what if a container has 3 replicas from alive nodes:

IN_SERVICE
IN_SERVICE
IN_MAINTENANCE

Then the IN_MAINTENANCE node gets shutdown, so there are only the 2 IN_SERVICE nodes left. This is not considered under-replicated and RM will take no action on this container, as it believes that the IN_MAINTENANCE copy will return.

SCM keeps track of replicas that are for Maintenance nodes and counts them as still being available so it can make decisions on this sort of thing.

The concern I have is that Recon shows different counts for under-replicated than the RM report and it can cause confusion to users.

xBis7 replied:
Like you said, for SCM, healthy replicas are also the ones that are offline [getHealthyReplicaCount()](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/RatisContainerReplicaCount.java#L193-L196).

When the SCM checks for over-replication or under-replication, it gets only the available replicas [getAvailableReplicas()](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/RatisContainerReplicaCount.java#L215-L221) based mostly on the `healthyReplicaCount` field.

The `healthyReplicaCount` doesn't count replicas from nodes in decommission or maintenance [countReplicas()](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/RatisContainerReplicaCount.java#L171).

```
this.healthyReplicas = healthyReplicas
```

I know that it says `healthyReplicas` but it should hold just the available replicas in order to be consistent with the SCM.

xBis7 replied:
As far as I understand, a node doesn’t go offline until its replicas have been copied to another node. While ENTERING_MAINTENANCE or DECOMMISSIONING container replicas are added or removed as needed to maintain proper replication. The container will be under-replicated until copies have been made and the node successfully becomes offline.

Once that is done, the container is correctly replicated, has 3 healthy and available replicas and 1 offline. SCM doesn’t report any under-replicated or over-replicated containers but Recon 

- for master, counts 1 over-replicated because it sees 4 replicas (no distinction between online - offline).
- for this patch, 0 count.

When the offline datanode is stopped, SCM doesn’t count unhealthy containers and

- for master, Recon no longer counts 1 over-replicated container.
- for this patch, no change in Recon.

sodonnel commented:
Yea, it is going to be tricky to fix this in Recon, without duplicating much of the logic in SCM for this scenario. In SCM there is some logic to avoid removing replicas on the dead node event when the replica is in maintenance, and then RM sees it as still present.

xBis7 commented:
> Yea, it is going to be tricky to fix this in Recon, without duplicating much of the logic in SCM for this scenario.

I was looking into that, but right now I think the best solution would be to reuse the `RatisContainerReplicaCount` class from SCM and let Recon get `under-replication` or `over-replication` from there.

For example, `ContainerHealthStatus#isOverReplicated()`

```diff
  public boolean isOverReplicated() {
-    return replicaDelta < 0;
+    return ratisContainerReplicaCount.isOverReplicated();
  }

```

No pending replica operations will be considered for Recon but it will be almost in sync with the SCM and any SCM changes for calculating `over-replication` or `under-replication` will be carried over to Recon without further changes. 

@sodonnel What do you think?

xBis7 commented:
While trying to make Recon consistent with the SCM, I ended up duplicating the code. I've updated `ContainerHealthStatus` to use `RatisContainerReplicaCount` for calculating over-replication or under-replication. This is the same class making these calculations in the SCM. With this approach, every change in how the SCM checks for over or under replication will be carried over to Recon without further modifications. 

I've also made a few other small changes to eliminate discrepancies with the SCM. For example, a container is missing when it has 0 replicas in the system, regardless of healthy or unhealthy replicas. Recon was marking a container as missing if it had 0 `healthy` replicas.

I've updated the existing tests and they are passing and I've also added new unit tests. I've also got a couple green CIs on my fork, so there isn't anything broken with these changes.

This approach is more extensive than the original one but it ensures that Recon stays in sync with SCM. @devmadhuu @sumitagrawl @sodonnel @adoroszlai Can you please take a look?  

There could be a follow up PR for making more methods in `ContainerHealthStatus` call their equivalents from `RatisContainerReplicaCount` and cleaning up the tests.

devmadhuu reviewed (commented):
And we had this JIRA [HDDS-9633](https://issues.apache.org/jira/browse/HDDS-9633) to sync Recon container stats with RM container report. Is this PR covering all ?

Can you add test matching admin CLI RM container report numbers matching for various states with Recon showing container stats or matching. If not showing on UI, at least does recon logging in logs ?

xBis7 commented:
@devmadhuu Thanks for the comments.

> isMissing() still considers total replicas irrespective of healthy/unhealthy. Shouldn't we consider only HealthyReplicas here ?

I changed that because I think it's a discrepancy. Missing container for the SCM means an empty replica list, regardless of healthy/unhealthy. If a container has only 1 replica and that replica is unhealthy, it won't be considered missing for SCM but Recon would be reporting it as such. 
I got that from these places in the code
* that's where `ReplicationManager` marks a container missing [here](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/ClosingContainerHandler.java#L74-L78), the container replica list has ALL known replicas. 
* that's where `LegacyReplicationManager` marks a container missing [here](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/LegacyReplicationManager.java#L535-L539) and [here](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/LegacyReplicationManager.java#L1834-L1837). 

Please correct me if I misunderstood and the list checked contain only healthy containers.

> And we had this JIRA [HDDS-9633](https://issues.apache.org/jira/browse/HDDS-9633) to sync Recon container stats with RM container report.

I wasn't aware of that. Thanks.

> Is this PR covering all ?

This is covering `over-replication`, `under-replication` and `missing`. I think we can get rid of `replicaDelta` in Recon entirely and reuse SCM code for all calculations. I didn't want to move ahead with it because the scope will grow out of hand. We will have to modify all the Recon tests for containers.

I also haven't change anything about `mis-replication`. I have a PR open for an issue related to that.

> Can you add test matching admin CLI RM container report numbers matching for various states with Recon showing container stats or matching.

Yes, let me look into an integration test or a robot test. I don't know which is easier at the moment. We should be able to use `curl` to get the data from Recon UI.

xBis7 commented:
@devmadhuu https://issues.apache.org/jira/browse/HDDS-9633 That was our initial issue and while investigating we narrowed it down to Recon ignoring out-of-service nodes.

xBis7 commented:
> However, what about EC containers? They will have the same problems, and there is an EcContainerReplicaCount class too. I see in getMinimumRequiredReplicaNum you are checking for EC or Ratis, so you also need to check for EC or Ratis and create the correct ReplicaCount object for EC or Ratis.

@sodonnel That's a valid point, let me look further into it. The way Recon is designed right now, it doesn't check for EC in any place and all the calculations are for Ratis. Also, I think there is no under-replication concept for EC, right? 

> Another question - what does Recon do to the container replicas it knows about when a maintenance node goes offline? Does it purge them from memory, or keep them? 

> What does recon do when an IN_SERVICE node goes dead, does it keep the replicas in memory marked as dead or does it just remove them like SCM does?

Both scenarios, dead node and a maintenance node going offline, are handled in the `DeadNodeHandler` and `ReconDeadNodeHandler` accordingly.

SCM doesn't remove replicas belonging to maintenance nodes, as I see [here](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/DeadNodeHandler.java#L88). Recon doesn't care about maintenance or the node state in general. `ReconDeadNodeHandler` triggers a health check which gets the dead node's containers from ReconSCM and processes them. If the container is over or under replicated, it will be added in the unhealthy container records. If the container is marked as DELETED in the SCM, Recon will mark it as DELETED in its container map. There are no other actions, therefore the replicas are kept in memory.

xBis7 commented:
@sodonnel I've updated the patch for EC. BTW, I'm working on an integration test like @devmadhuu suggested.

xBis7 commented:
While working on the integration test, I created an EC container and then compared SCM to Recon. I found out that Recon-SCM doesn't store EC containers. I filed this jira for the issue

https://issues.apache.org/jira/browse/HDDS-9898

sodonnel commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ContainerReplicaCount.java:
```diff
@@ -49,6 +49,8 @@ public interface ContainerReplicaCount {
   boolean isSufficientlyReplicatedForOffline(DatanodeDetails datanode,
       NodeManager nodeManager);
 
+  boolean isUnderReplicated();
```
I am not sure we really need another method on the interface which is just `!isSufficientlyReplicated()` - we can just call `!isSufficientlyReplicated()` without adding more to the interface.

sodonnel commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthStatus.java:
```diff
@@ -78,6 +100,14 @@ public int getReplicationFactor() {
   }
 
   public boolean isHealthy() {
+    return containerReplicaCount.isHealthy();
```
This is changing the meaning here. RepCount isHealthy is checking for the container being closed / quasi-closed and all replicas match that. The original isHealthy here was about not under / over rep and also not mis-rep.

xBis7 replied:
I know, that's why I renamed the original method to `isProperlyReplicated()` and now every method in Recon, calling `isHealthy()`, now calls `isProperlyReplicated()`.

It's part of the problem, that Recon uses the same method naming as the SCM but has a different meaning for most of them.

xBis7 commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthStatus.java:
```diff
@@ -87,19 +117,19 @@ public boolean isDeleted() {
 
   public boolean isUnderReplicated() {
-    return !isMissing() && replicaDelta > 0;
+    return !isMissing() && containerReplicaCount.isUnderReplicated();
```
@sodonnel By removing `isUnderReplicated()` from the interface, this should be
```suggestion
    return !isMissing() && !containerReplicaCount.isSufficientlyReplicated();
```
right?

xBis7 commented:
@devmadhuu I've added the integration test.

To make sure that the test isn't flaky, I created a workflow in another branch, that runs the test on repeat. Each workflow runs the test 10 times. I got 3 green CIs.

https://github.com/xBis7/ozone/actions/runs/7197563498/job/19604976469#step:6:4035
https://github.com/xBis7/ozone/actions/runs/7197941719/job/19606235980#step:6:4035
https://github.com/xBis7/ozone/actions/runs/7197993201/job/19606410429#step:6:4035

devmadhuu commented:
> @devmadhuu Thanks for the comments.
> 
> > isMissing() still considers total replicas irrespective of healthy/unhealthy. Shouldn't we consider only HealthyReplicas here ?
> 
> I changed that because I think it's a discrepancy. Missing container for the SCM means an empty replica list, regardless of healthy/unhealthy. If a container has only 1 replica and that replica is unhealthy, it won't be considered missing for SCM but Recon would be reporting it as such. I got that from these places in the code
> 
>     * `ReplicationManager` marks a container missing [here](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/ClosingContainerHandler.java#L74-L78), the container replica list has ALL known replicas.
> 
>     * `LegacyReplicationManager` marks a container missing [here](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/LegacyReplicationManager.java#L535-L539) and [here](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/LegacyReplicationManager.java#L1834-L1837).
> 
> 
> Please correct me if I misunderstood and the checked list contain only healthy containers.
> 
> > And we had this JIRA [HDDS-9633](https://issues.apache.org/jira/browse/HDDS-9633) to sync Recon container stats with RM container report.
> 
> I wasn't aware of that. Thanks.
> 
> > Is this PR covering all ?
> 
> This is covering `over-replication`, `under-replication` and `missing`. I think we can get rid of `replicaDelta` in Recon entirely and reuse SCM code for all calculations. I didn't want to move ahead with it because the scope will grow out of hand. We will have to modify all the Recon tests for containers.
> 
> I also haven't changed anything about `mis-replication`. I have a PR open for an issue related to that.
> 
> > Can you add test matching admin CLI RM container report numbers matching for various states with Recon showing container stats or matching.
> 
> Yes, let me look into an integration test or a robot test. I don't know which is easier at the moment. We should be able to use `curl` to get the data from Recon UI.

@xBis7 thanks for continue working on this, As I understand Recon also was considering the container as missing when all 3 replicas are zero, so both SCM and Recon were treating missing container as same. You check the original code of Recon ContainerHealthStatus#isMissing method where Recon was using numReplicas == 0 match..

devmadhuu commented:
@xBis7 there is too much of discussion in this pR, it would be helpful if you summarize the final changes in a para here ?what changed in Recon compared to SCM ?

xBis7 commented:
@devmadhuu Here is a summary of the changes in this PR:

Recon and SCM are using the same naming and terminology regarding unhealthy containers. The code in Recon was quite old and following a very simple process of identifying each kind of unhealthy container, compared to the SCM. Recon was ignoring the existence of offline nodes (decommission, maintenance) and was also providing a different explanation for some terminologies.

The goal of this PR, is to make Recon container page consistent with the ReplicationManager `admin container report`. To avoid duplicating the code or its logic, I reused the same methods that the ReplicationManager is using for identifying under-replication, over-replication etc. With these changes, Recon is using the same algorithms as the RM for unhealthy containers. Every change in the RM is carried over to Recon as well. 

I think it's important to deal with common terminology, in the same manner, to be consistent and avoid confusion. I'm referring to the way Recon defines healthy and missing containers.

* healthy
  * RM
    *  the container is closed or quasi-closed and all replicas are in the same state
  * Recon
    *  the container isn't over-replicated or under-replicated or mis-replicated 
    * I didn't change the code but I renamed the method and all of its references. This is actually referring to proper replication and not what RM means by healthy.
* missing
  * RM
    * 0 replicas, no information or reference at all. An `admin container info` call from the CLI, returns empty for all fields. If a container has only 1 unhealthy replica and we can get at least some info from it, like pipeline ID, then the container isn't missing.
  * Recon
    * `ContainerHealthStatus` was getting a list of all the replicas and then filtering it to get just the healthy ones. Then checking the healthy replica list and if there are no healthy replicas, then the container is missing.
    * I modified it to use the initial replica list, for checking for missing.

We can revert the healthy and missing changes to reduce the scope of this PR.

I've added an integration test, that places some nodes in decommission or maintenance and makes a container go missing. The test, compares the RM `admin container report` to Recon's response in every step and makes sure that the numbers are the same and that RM's sample of containers exist in Recon's list of unhealthy containers.

devmadhuu commented:
> @devmadhuu Here is a summary of the changes in this PR:
> 
> Recon and SCM are using the same naming and terminology regarding unhealthy containers. The code in Recon was quite old and following a very simple process of identifying each kind of unhealthy container, compared to the SCM. Recon was ignoring the existence of offline nodes (decommission, maintenance) and was also providing a different explanation for some terminologies.
> 
> The goal of this PR, is to make Recon container page consistent with the ReplicationManager `admin container report`. To avoid duplicating the code or its logic, I reused the same methods that the ReplicationManager is using for identifying under-replication, over-replication etc. With these changes, Recon is using the same algorithms as the RM for unhealthy containers. Every change in the RM is carried over to Recon as well.
> 
> I think it's important to deal with common terminology, in the same manner, to be consistent and avoid confusion. I'm referring to the way Recon defines healthy and missing containers.
> 
>     * healthy
>       
>       * RM
>         
>         * the container is closed or quasi-closed and all replicas are in the same state
>       * Recon
>         
>         * the container isn't over-replicated or under-replicated or mis-replicated
>         * I didn't change the code but I renamed the method and all of its references. This is actually referring to proper replication and not what RM means by healthy.
> 
>     * missing
>       
>       * RM
>         
>         * 0 replicas, no information or reference at all. An `admin container info` call from the CLI, returns empty for all fields. If a container has only 1 unhealthy replica and we can get at least some info from it, like pipeline ID, then the container isn't missing.
>       * Recon
>         
>         * `ContainerHealthStatus` was getting a list of all the replicas and then filtering it to get just the healthy ones. Then checking the healthy replica list and if there are no healthy replicas, then the container is missing.
>         * I modified it to use the initial replica list, healthy and unhealthy.
> 
> 
> We can revert the healthy and missing changes to reduce the scope of this PR.
> 
> I've added an integration test, that places some nodes in decommission or maintenance and makes a container go missing. The test, compares the RM `admin container report` to Recon's response in every step and makes sure that the numbers are the same and that RM's sample of containers exist in Recon's list of unhealthy containers.

Thanks @xBis7 for your effort on this patch. I think this looks much better in current state and regarding EC handling of containers, this can be taken up separately.

sumitagrawl commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthStatus.java:
```diff
@@ -78,6 +100,14 @@ public int getReplicationFactor() {
+    return containerReplicaCount.isSufficientlyReplicated();
+  }
+
+  public boolean isProperlyReplicated() {
```
isSufficientlyReplicated is where delta is 0 or over-replication, and isProperlyReplicated is delta 0 (no under/over replication) and no misReplication.
I think over-replication can be considered as healthy only (may not need check delta=0).
When perfectly replicated (delta=0), in that case need check for misReplication

We can rename method as isHealtyReplication() with above 2 condition, as sufficently replication also means properly replication.

Can see some comment in code related to this point,
org.apache.hadoop.hdds.scm.SCMCommonPlacementPolicy#validateContainerPlacement

> // There are scenarios where there could be excessive replicas on a rack due to nodes
>     // in decommission or maintenance or over-replication in general.
>     // In these cases, ReplicationManager shouldn't report mis-replication.
>     // The original intention of mis-replication was to indicate that there isn't enough rack tolerance.
>     // In case of over-replication, this isn't an issue.
>     // Mis-replication should be an issue once over-replication is fixed, not before.
>     // Adjust the maximum number of replicas per rack to allow containers
>     // with excessive replicas to not be reported as mis-replicated.
>     maxReplicasPerRack += Math.max(0, dns.size() - replicas);
>     return new ContainerPlacementStatusDefault(
>         currentRackCount.size(), requiredRacks, numRacks, maxReplicasPerRack,
>             currentRackCount);

xBis7 replied:
@sumitagrawl I couldn't figure out how to call this method. The original name was `isHealthy` and I just wanted to change that, to avoid having a different meaning here than the SCM has.

> We can rename method as isHealtyReplication() with above 2 condition,

That makes sense to me. I'll rename it.

 > Can see some comment in code related to this point,
org.apache.hadoop.hdds.scm.SCMCommonPlacementPolicy#validateContainerPlacement

I added that comment in a previous PR, to explain that when having `over-replication`, we shouldn't also have `mis-replication`. Although, `over-replication` is an unhealthy state, `mis-replication` refers to not having enough racks or having an uneven replica distribution per rack. Having too many replicas in a rack, isn't an issue of `mis-replication` but just `over-replication`.

> I think over-replication can be considered as healthy only (may not need check delta=0).

With the above comment, it doesn't mean that when a container is over-replicated then it's entirely healthy, we still need to fix over-replication(disk space issue, etc.).

xBis7 replied:
Long-term, the best approach would be to remove all delta calculations from this class and rely entirely on `RatisContainerReplicaCount` and `ECContainerReplicaCount`.

xBis7 commented:
@sodonnel @sumitagrawl Can you take another look on this PR? Any comments on the approach?

devmadhuu commented:
@xBis7 - Thanks for the addressing the comments. Any comments left ? What is pending ?

xBis7 commented:
@devmadhuu I deliberately changed the meaning in some of Recon's methods, to make it consistent with the ReplicationManager. If @sodonnel and @sumitagrawl don't have an objection to the changes, then there isn't something pending.

sodonnel commented:
If @devmadhuu is happy I am happy for this to be committed.

sumitagrawl commented:
>> I think over-replication can be considered as healthy only (may not need check delta=0).

> With the above comment, it doesn't mean that when a container is over-replicated then it's entirely healthy, we still need to > fix over-replication(disk space issue, etc.).

This is only difference wrt org.apache.hadoop.hdds.scm.SCMCommonPlacementPolicy#validateContainerPlacement having description of the behavior. https://github.com/apache/ozone/pull/5651#discussion_r1441647606

@sodonnel what is your opinion for this? over-replication IMO does not make container as unhealty.

xBis7 commented:
@sumitagrawl Healthy has a different meaning for the ReplicationManager and for Recon and so it's tricky to define it. For RM, an over-replicated container isn't unhealthy. But Recon defines all states where we don't have the correct number of replicas as unhealthy, so that the containers can appear in the container-endpoint.

If I understand correctly, you are suggesting to change this 
```java
  public boolean isProperlyReplicated() {
    return replicaDelta == 0 && !isMisReplicated();
  }
```
to this
```java
  public boolean isProperlyReplicated() {
    return replicaDelta >= 0 && !isMisReplicated();
  }
```
Is that correct?

So that we can say that a container is properly replicated if it's neither under-replicated nor mis-replicated.

sumitagrawl commented:
> So that we can say that a container is properly replicated if it's neither under-replicated nor mis-replicated.

ok, You mean those container will be part of below category in Recon UI, right?
![image](https://github.com/apache/ozone/assets/23372859/97d30c67-69c2-49a9-897e-8036715cef4d)


If so, I am fine with the change.

xBis7 commented:
@sumitagrawl Let me explain

> But Recon defines all states where we don't have the correct number of replicas as unhealthy, so that the containers can appear in the container-endpoint.

Healthy has a different meaning for Recon. Recon recognizes 5 states
1. Over-replication
2. Under-replication
3. Mis-replication
4. Missing
5. Correct number of replicas

The comment from the `SCMCommonPlacementPolicy` says that we shouldn't have to worry about over-replication because none of the above states means unhealthy for SCM. This is what an unhealthy container is for SCM, check [here](https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ContainerReplicaCount.java#L58-L74).

 Recon says that if the container is in State 5 then it's healthy and if it's in any other state, I'll present it in the UI. The meaning here is different. If we change it the scope of the changes will grow out of hand.

> > So that we can say that a container is properly replicated if it's neither under-replicated nor mis-replicated.
>
> ok, You mean those container will be part of below category in Recon UI, right?

With this change, over-replicated containers won't appear in the UI. This is from master, if an `over-replicated` container is considered healthy, then this check [here](https://github.com/apache/ozone/blob/master/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthTask.java#L316-L318) will return and the container will never be added in the unhealthy containers record [here](https://github.com/apache/ozone/blob/master/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthTask.java#L323). 

We need the container to be added in the unhealthy containers record, because it's used for the UI. Check [here](https://github.com/apache/ozone/blob/master/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthTask.java#L423) which is where we distinguish over-replication. In that method, it first checks if the container is healthy [here](https://github.com/apache/ozone/blob/master/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthTask.java#L429) and then checks for over-replication [here](https://github.com/apache/ozone/blob/master/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthTask.java#L475).

adoroszlai commented:
Thanks @xBis7 for the patch, @devmadhuu, @sodonnel, @sumitagrawl for the review.

I left a few minor suggestions for future changes.
